### PR TITLE
Add support for BME280 temp/humidity/pressure sensor

### PR DESCRIPTION
**Pull request by biotinker**
Not sure who else should be on this PR?

### CODE REVIEW

**Review by randhid — CHANGES_REQUESTED** [review]
I'll do another pass, but we're not looking to add more components to rdk now. Scope doc for eliot please.

**Comment by biotinker** [comment]
Wasn't making stuff like this part of the point of hack week? So that we can have better support for things, in addition to finding existing bugs?

I can make a scope doc, but surely there's some better way to add support for individual common sensor components than writing a document for the CEO first for each one?

**Comment by biotinker** [comment]
@erh

**Review by erh — APPROVED** [review]
i'm good with things like this.
it's not an API change.

File: components/sensor/bme280/bme280.go [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+	if err != nil {
+		s.logger.Debug("Failed to request temperature")
+	}
+	buffer, err := handle.Read(ctx, 8)
+	if err != nil {
+		return nil, err
+	}
+	if len(buffer) != 8 {
+		return nil, errors.New("i2c read did not get 8 bytes")
+	}
```
Comment (randhid) [inline components/sensor/bme280/bme280.go:224]: Again with the design of this sensor. You're sure you want to read then close the handle every-time we call Readings?

**Review by randhid — CHANGES_REQUESTED** [review]

File: components/sensor/bme280/bme280.go [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+	err = s.setOverSample(ctx, bme280CTRLMEASReg, 5, 1) // Default of 1x oversample
+	if err != nil {
+		return nil, err
+	}
+	// pressure
+	err = s.setOverSample(ctx, bme280CTRLMEASReg, 2, 1) // Default of 1x oversample
+	if err != nil {
+		return nil, err
+	}
+
```
[the remark this answers is not in the document]
Comment (biotinker) [reply components/sensor/bme280/bme280.go:190]: What would be the use of polling the sensor continuously if nothing is consuming the reported data? That would just be useless noise on the i2c bus, no?

The commands above will tell the sensor to be continuously running, so up-to-date values are available immediately as needed.

File: components/sensor/bme280/bme280.go (hunk 1 above)
Comment (biotinker) [reply components/sensor/bme280/bme280.go:224]: Multiple devices share the same i2c bus, so we can't just keep the handle open indefinitely.

File: components/sensor/bme280/bme280.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/sensor/bme280/bme280.go:190]: It depends on the sensor itself, if this is something you want to poll continuously, like an imu or encoder, then this isn't the design to poll it.

If it is not, then this is the design.

File: components/sensor/bme280/bme280.go (hunk 1 above)
Comment (randhid) [reply components/sensor/bme280/bme280.go:224]: Not what I was saying, you'd open/close in each loop if you poll. But it seems like you don't want to poll it continuously.

**Comment by randhid** [comment]
Please let vision know about the flaky test.

File: components/sensor/bme280/bme280.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply components/sensor/bme280/bme280.go:190]: Yeah, almost certainly not. So this is the design.

File: components/sensor/bme280/bme280.go [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+func newSensor(
+	ctx context.Context,
+	deps registry.Dependencies,
+	name string,
+	attr *AttrConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+	b, err := board.FromDependencies(deps, attr.Board)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply components/sensor/bme280/bme280.go:120]: Confirmed:

```
1.6648216731985579e+09  error   robot_server    server/entrypoint.go:124        Fatal error running server, existing now: error validating "components.1": "i2c bus" is required
2022-10-03T19:27:53.310+0100    FATAL   robot_server    utils@v0.0.6-0.20220920201121-47f4f3281a18/runtime.go:53        error validating "components.1": "i2c bus" is required
```

**Review by randhid — APPROVED** [review]
Let's Get Temperatures and MilliPascals

File: components/sensor/bme280/bme280.go [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+	rhs += 5.02808 * math.Log10(ratio)
+	rhs += -1.3816e-7 * (math.Pow(10, (11.344*(1-1/ratio))) - 1)
+	rhs += 8.1328e-3 * (math.Pow(10, (-3.49149*(ratio-1))) - 1)
+	rhs += math.Log10(1013.246)
+
+	// factor -3 is to adjust units - Vapor Pressure SVP * humidity
+	vp := math.Pow(10, rhs-3) * humid
+	// (2) DEWPOINT = F(Vapor Pressure)
+	t := math.Log(vp / 0.61078) // temp var
+	return (241.88 * t) / (17.558 - t)
```
Comment (randhid) [inline components/sensor/bme280/bme280.go:319]: Not blocking but probably would be good to check NaN condition somewhere.

Comment (biotinker) [reply components/sensor/bme280/bme280.go:319]: In this particular case it's not mathematically possible for this to actually be NaN due to imitations on the byte widths of the inputs to temp and humidity. Nevertheless I will add a check.

**Review by mcvella — APPROVED** [review]

**Pull request merged**